### DISCUSSION TITLE
Slower word-based movements as they trip over punctuation

### DISCUSSION THREAD

**Discussion by eugenesvk** [helix-editor/helix#5590] [General]
I've noticed that word-based movements are slower in Helix as it seems to treat punctuation chars as separate words

For example, on this contrived text it takes me twice the number of keystrokes to jump to the end (`^` signify cursor positions after each command)
```
word-li1;word-li2
    ^   ^    ^   ^ Sublime Text's `word_ends`
    ^   ^    ^   ^ emacs' `forward-word`
   ^^  ^^   ^^  ^ Helix's `move_next_word_end`
```

Could a setting be added that would allow the command to skip those punctuation and treat only alphanumeric chars as words?

**Comment by archseer** [comment]
A lot of care has been put in to ensure we match kakoune's word movements: https://github.com/helix-editor/helix/blob/8d273a561387ceb0bcc9c9f8c9fa90a64204118f/helix-core/src/movement.rs#L528-L1127

In particular the difference in behavior here is that vim will position your cursor right after the word, but kakoune/helix instead **selects** the word and stops on the last character. Following vim behavior isn't desirable here since keymaps like `wd` would now be broken (since they'd also delete the following character). For longer jumps you could use `W`? Or `f`

**Comment by eugenesvk** [comment]
> kakoune/helix instead **selects** the word and stops on the last character

that's exactly the same behavior with word movements after `set-mark-command` is enabled in emacs: it stops on the last character of the first word `word`, so Helix's `wd` wouldn't be broken, but then it still doesn't treat `-` as a separate word, it selects `-li1` on the second jump while Helix selects `-` on the second jump and `lil1` only on the 3rd jump

Will something else be horribly broken if punctuation is glued to the beginning of the "next word", but not to end of "this word" [^1]?

> For longer jumps you could use W?

Yes, but I'd prefer my shorter jumps to work since longer ones are behind a modifier, so much less convenient
But more importantly, I'd actually prefer the punctuation marks to be part of some word because then I wouldn't have this weird behavior when holding my forward-delete-by-word command `["move_next_word_end","delete_selection"] # ⌦w` leaves punctuation marks on otherwise empty lines
It's perfectly fine if punctuation isn't part of the first word, here I'd agree that it's better to select only a word `word` instead of `word-`, but then I don't see a need to select only a punctuation mark, and if I need to get rid of it, I'd just go word back — the punctuation mark would disappear from selection and I'd only select `li1` instead of `-li1`

[^1]: by the way, I found the Helix command names a bit confusing: if I'm at a the beginning of `wfirst-wsecond`, the command `move_next_word_end` actually moves to the end of __THIS__ word `wfirst`, not the next one, the next word is `wsecond`
